In [4]:
import torch
import numpy as np

# Bài 1: Tạo Tensor cơ bản

In [5]:
import torch
a= torch.tensor([[1,2,3],
                 [4,5,6]],
                 dtype=torch.int32
                 )
print("tensor a: ")
print(a)
print("kich thuoc: ", a.shape)
print("so chieu: ", a.ndim)
print("kieu du lieu: ", a.dtype)


tensor a: 
tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
kich thuoc:  torch.Size([2, 3])
so chieu:  2
kieu du lieu:  torch.int32


## Nhận xét
 - tensor a có kích thước 2x3
 - tensor a có 2 chiều
 - kiểu dữ liệu của tensor a là torch.int32


# Bài 2: Tạo các Tensor thường dùng

In [6]:
a= torch.rand(3,4)
b= torch.randn(3,4)
c= torch.zeros( 2,3)
d= torch.ones(2,3)
e= torch.eye(4)

print(a)
print(b)
print(c)
print(d)
print(e)

tensor([[0.8220, 0.5050, 0.1746, 0.5286],
        [0.5065, 0.1866, 0.5572, 0.5011],
        [0.3438, 0.4783, 0.2708, 0.6619]])
tensor([[-0.4211,  0.3042, -0.0294,  0.2832],
        [ 0.3965, -0.9886, -0.7212, -0.0896],
        [-0.5296,  1.3613, -1.3067, -1.8140]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


In [7]:

f= torch.full((5,5), 10)
print(f)

tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


## Nhận xét
- torch.rand() tạo số ngẫu nhiên theo phân phối đều từ 0 đến 1
- torch.randn() tạo số ngẫu nhiên theo phân phối chuẩn
- torch.zeros() tạo tensor toàn số 0
- torch.ones() tạo tensor toàn số 1
- torch.eye() tạo ma trận đơn vị
- torch.full((5,5), 10) tạo tensor 5x5 chứa toàn giá trị 10 

# Bài 3: Kiểm tra thông tin tensor

In [8]:
a= torch.zeros(2,3,4)
print("tensor: ")
print(a)
print("shape: ", a.shape)
print("size: ", a.size())
print("number of dimesions: ", a.ndim)
print("data type: ", a.dtype)


tensor: 
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
shape:  torch.Size([2, 3, 4])
size:  torch.Size([2, 3, 4])
number of dimesions:  3
data type:  torch.float32


## Nhận xét
- torch.size([2,3,4]) cho biết tensor có 3 chiều với kích thước lần lượt là 2,3,4

# Bài 4: Slicing và Indexing

In [9]:
a= torch.rand(3,4,5)
print("original tensor: ")
print(a)

print("\ntensor a[1]: ")
print(a[1])

print("\ntensor a[1:, 2:4] :")
print(a[1:, 2:4])

original tensor: 
tensor([[[0.8353, 0.6635, 0.8383, 0.0774, 0.5515],
         [0.1351, 0.2790, 0.2289, 0.8467, 0.3784],
         [0.0230, 0.8221, 0.8016, 0.0928, 0.1873],
         [0.0241, 0.6731, 0.4063, 0.7285, 0.1502]],

        [[0.1099, 0.4854, 0.9786, 0.2696, 0.2230],
         [0.9188, 0.2382, 0.8214, 0.4642, 0.7156],
         [0.5452, 0.5151, 0.9338, 0.3660, 0.3564],
         [0.1679, 0.8966, 0.4672, 0.7279, 0.1402]],

        [[0.6790, 0.4691, 0.7245, 0.7462, 0.6009],
         [0.1806, 0.4370, 0.3647, 0.9585, 0.9483],
         [0.7649, 0.6219, 0.2999, 0.2395, 0.0771],
         [0.7285, 0.7710, 0.8435, 0.0874, 0.6592]]])

tensor a[1]: 
tensor([[0.1099, 0.4854, 0.9786, 0.2696, 0.2230],
        [0.9188, 0.2382, 0.8214, 0.4642, 0.7156],
        [0.5452, 0.5151, 0.9338, 0.3660, 0.3564],
        [0.1679, 0.8966, 0.4672, 0.7279, 0.1402]])

tensor a[1:, 2:4] :
tensor([[[0.5452, 0.5151, 0.9338, 0.3660, 0.3564],
         [0.1679, 0.8966, 0.4672, 0.7279, 0.1402]],

        [[0.7649, 0.621

In [10]:
print("lay 2 phan tu cuoi tren chieu thu 3: ")
print(a[: , : , -2])

lay 2 phan tu cuoi tren chieu thu 3: 
tensor([[0.0774, 0.8467, 0.0928, 0.7285],
        [0.2696, 0.4642, 0.3660, 0.7279],
        [0.7462, 0.9585, 0.2395, 0.0874]])


## Nhận xét
- a có shape là torch.size([3,4,5])
- a[1] có shape là torch.size([4,5])
- a[1:, 2:4] có shape là torch.size([2,2,5])
- a[:, :, -2] là lấy 2 phần tử cuối trên chiều thứ 3

# Bài 5: Reshape và Flatten

In [11]:
a= torch.rand(3,4)
print("original: ")
print(a)

print("\nFlatten: ")
print(a.ravel())

print("\nReshape to 3x2x2: ")
print(a.reshape(3,2,2))

print("\nReshape thanh kich thuoc 2x6: ")
print(a.reshape(2,6))

print("\nSo phan tu ban dau: ", a.numel())
print("\nSo phan tu sau khi reshape 3x2x2: ", a.reshape(3,2,2).numel())
print("\nSo phan tu sau khi reshape 2x6: ", a.reshape(2,6).numel())

original: 
tensor([[0.0303, 0.1035, 0.6783, 0.6028],
        [0.2161, 0.2818, 0.5813, 0.7641],
        [0.1781, 0.8062, 0.1744, 0.2027]])

Flatten: 
tensor([0.0303, 0.1035, 0.6783, 0.6028, 0.2161, 0.2818, 0.5813, 0.7641, 0.1781,
        0.8062, 0.1744, 0.2027])

Reshape to 3x2x2: 
tensor([[[0.0303, 0.1035],
         [0.6783, 0.6028]],

        [[0.2161, 0.2818],
         [0.5813, 0.7641]],

        [[0.1781, 0.8062],
         [0.1744, 0.2027]]])

Reshape thanh kich thuoc 2x6: 
tensor([[0.0303, 0.1035, 0.6783, 0.6028, 0.2161, 0.2818],
        [0.5813, 0.7641, 0.1781, 0.8062, 0.1744, 0.2027]])

So phan tu ban dau:  12

So phan tu sau khi reshape 3x2x2:  12

So phan tu sau khi reshape 2x6:  12


## Nhận xét:
- tensor ban đầu 3x4 là 12
- sau khi reshape 3x2x2 là 12
- sau khi reshape 2x6 là 12
- vậy reshape chỉ thay đổi hình dạng của tensor chứ không thay đổi tổng số phần tử

# Bài 6: Phép toán và nhân ma trận

In [12]:
a= torch.randn(2,3)
b= torch.randn(2,3)
print("a: ")
print(a)
print("\nb: ")
print(b)
print("\na + b: ")
print(a+b)
print("\na / b: ")
print(a/b)
print("\na^2: ")
print(a**2)
print("\nmatrix multiplication( a@b.T): ")
print(a @ b.T)

print("\nshape cua a@b.T : ")
print((a@b.T).shape)

a: 
tensor([[ 1.0655, -1.4657, -1.1212],
        [-1.1814, -0.0871,  2.9330]])

b: 
tensor([[-0.8390,  1.1311, -0.7689],
        [-0.2776, -0.1777,  1.0280]])

a + b: 
tensor([[ 0.2265, -0.3346, -1.8901],
        [-1.4590, -0.2648,  3.9611]])

a / b: 
tensor([[-1.2700, -1.2958,  1.4581],
        [ 4.2565,  0.4904,  2.8531]])

a^2: 
tensor([[1.1353e+00, 2.1482e+00, 1.2570e+00],
        [1.3957e+00, 7.5948e-03, 8.6027e+00]])

matrix multiplication( a@b.T): 
tensor([[-1.6896, -1.1879],
        [-1.3627,  3.3586]])

shape cua a@b.T : 
torch.Size([2, 2])


## Nhận xét:
- phép toán a*b là nhân từng phần tử tương ứng
- phép toán a@b.T là nhân ma trận 
- với a có kích thước 2x3 còn b.T có kích thước 3x2, nên shape của kết quả a@b.T là 2x2


# Bài 7: Các hàm thống kê trên Tensor

In [13]:
a= torch.randn(3,4)
print(a)
print("\nMean: ")
print(torch.mean(a, dim=0))

print("\nStandard deviation: ")
print(torch.std(a, dim=0))

print("\nCumulative sum: ")
print(torch.cumsum(a, dim=0))


print("\nMean: ")
print(torch.mean(a, dim=1))

print("\nStandard deviation: ")
print(torch.std(a, dim=1))

print("\nCumlative sum: ")
print(torch.cumsum(a, dim=1))

tensor([[-0.8937,  1.7144,  0.9326,  0.5376],
        [ 0.3355,  1.9705,  0.5185, -0.1979],
        [ 0.1650, -0.0589, -2.0253, -1.2554]])

Mean: 
tensor([-0.1311,  1.2087, -0.1914, -0.3052])

Standard deviation: 
tensor([0.6659, 1.1052, 1.6016, 0.9013])

Cumulative sum: 
tensor([[-0.8937,  1.7144,  0.9326,  0.5376],
        [-0.5582,  3.6849,  1.4511,  0.3397],
        [-0.3932,  3.6260, -0.5742, -0.9157]])

Mean: 
tensor([ 0.5727,  0.6566, -0.7937])

Standard deviation: 
tensor([1.0931, 0.9271, 1.0310])

Cumlative sum: 
tensor([[-0.8937,  0.8207,  1.7534,  2.2909],
        [ 0.3355,  2.3059,  2.8244,  2.6265],
        [ 0.1650,  0.1061, -1.9192, -3.1746]])


## Nhận xét:
- dim=0 thực hiện theo chiều đầu tiên nên kết quả thống kê theo từng cột
- dim=1 thực hiện theo chiều thứ 2 nên kết quả thống kê theo từng hàng
- khi đổi dim=0 sang dim=1, phép tính thực hiện theo chiều khác nhau nên kết quả chuyển từ thống kê theo cột sang thống kê theo hàng

# Bài 8: Autograd cơ bản

In [14]:
x= torch.tensor(
    3.6,
    requires_grad=True
)
y= x*x
y.backward()

print("x= ", x)
print("y= ", y)
print("x.grad= ", x.grad)

x=  tensor(3.6000, requires_grad=True)
y=  tensor(12.9600, grad_fn=<MulBackward0>)
x.grad=  tensor(7.2000)


## kiểm tra gradient
-  ta có: y = x²
- đạo hàm y'= 2x
- tại x=3.6 thì y'= 2 x 3.6 = 7.2
- kết quả bằng với x.grad 

## Đổi hàm  thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.

In [15]:
x= torch.tensor(
    3.6,
    requires_grad=True
)
y= 3*x**2+2*x+1
y.backward()

print("x= ", x)
print("y= ", y)
print("x.grad= ", x.grad)


x=  tensor(3.6000, requires_grad=True)
y=  tensor(47.0800, grad_fn=<AddBackward0>)
x.grad=  tensor(23.6000)


## Kiểm tra gradient
- ta có: y = 3x²+2x+1
- đạo hàm: y'= 6x+2
- tại x=3.6 thì y'= 6 x 3.6 + 2= 23.6

# Bài 9: Gradient Descent tìm hệ số đa thức
- Sinh dữ liệu từ đa thức y = x² + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [33]:
polynomial= np.poly1d([1,2,3]) #đa thức thật
N=20 # số lượng dữ liệu

X= np.random.randn(N,1)*5 #sinh dữ liệu X
Y= polynomial(X) #tính Y theo đa thức

XX= np.hstack([  # tạo ma trận đặc trưng [x^2, x, 1]
    X*X,
    X,
    np.ones_like(X)
])

w= torch.randn(3,1, requires_grad=True) #3 hệ số cần tìm

#chuyển dữ liệu sang tensor
x= torch.tensor(XX, dtype=torch.float32)
y= torch.tensor(Y, dtype=torch.float32)

optimizer=torch.optim.NAdam([w], lr=0.01)

print("initial coeficients: ")
print(w)

# huấn luyện
for epoch in range (1000):
    optimizer.zero_grad() # xoá gradient của lần lặp trước
    y_pred= x@w
    mse= torch.mean(torch.square(y-y_pred))
    mse.backward() # tính gradient của hàm mất mát mse đối với các hệ số
    optimizer.step() #dùng gradient để cập nhật các hệ số, làm cho sai số nhỏ hơn

print("learned coeficients: ")
print(w)

print("\nExpected coeficients: ")
print("[1,2,3]")



initial coeficients: 
tensor([[0.3744],
        [0.1866],
        [1.6633]], requires_grad=True)
learned coeficients: 
tensor([[1.0017],
        [2.0026],
        [2.6859]], requires_grad=True)

Expected coeficients: 
[1,2,3]


In [ ]:
print("\nSo sanh he so:")

print("He so tim duoc:")
print(w.detach().numpy())

print("\nHe so mong doi:")
print("[1, 2, 3]")


So sanh he so:
He so tim duoc:
[[1.0017318]
 [2.0026054]
 [2.685925 ]]

He so mong doi:
[1, 2, 3]


## So sánh 3 hệ số học được với [1,2,3]
- nhận được 3 hệ số gần với [1,2,3]
## Giải thích vai trò của  optimizer.zero_grad(), mse.backward() và optimizer.step(). 
- optimizer.zero_grad():  xoá gradient của lần lặp trước
- mse.backward(): tính gradient của hàm mất mát mse đối với các hệ số
- optimizer.step(): dùng gradient để cập nhật các hệ số nhằm làm cho sai số nhỏ hơn

# Bài 10: Giải hệ phương trình bằng Autograd 
- Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình. 

In [30]:
import random
A= torch.tensor(random.random(), requires_grad=True)
B= torch.tensor(random.random(), requires_grad=True)
C= torch.tensor(random.random(), requires_grad=True)
D= torch.tensor(random.random(), requires_grad=True)

EPOCHS =2000
optimizer= torch.optim.NAdam([A,B,C,D], lr=0.01)

for _ in range(EPOCHS):
   #sai số của phương trình
   y1= A+B-9
   y2= C-D-1
   y3= A+C-8
   y4= B-D-2

   #tổng bình phương sai số
   sqerr= y1 * y1 + y2 * y2 + y3 * y3 + y4 * y4

   optimizer.zero_grad()
   sqerr.backward()
   optimizer.step()

print("A= ", A)
print("B= ", B)
print("C= ", C)
print("D= ", D)



A=  tensor(4.4839, requires_grad=True)
B=  tensor(4.5160, requires_grad=True)
C=  tensor(3.5161, requires_grad=True)
D=  tensor(2.5160, requires_grad=True)


In [31]:
print("\nKiem tra he phuong trinh: ")
print("A + B= ", (A+B).item())
print("C - D= ", (C-D).item())
print("A + C= ", (A+C).item())
print("B - D= ", (B-D).item())


Kiem tra he phuong trinh: 
A + B=  8.999951362609863
C - D=  1.0000178813934326
A + C=  7.999981880187988
B - D=  1.9999876022338867


## Nhận xét:
- hệ phương trình có nhiều nghiệm vì có 4 ẩn nhưng các phương trình không độc lập để xác định duy nhất cả 4 ẩn. có thể chọn biến tự do, các biến còn lại được xác định theo biến đó

# Bài 11:  Viết chương trình PyTorch để thực hiện các yêu cầu sau: 
• Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4. 
• Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột. 
• Cho y = 2x² + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công. 
• Thay đa thức trong Bài 9 bằng y = 2x² - 3x + 5; huấn luyện và cho biết các hệ số tìm được. 